### PR TITLE
[SER-500] Deserialization, take 2

### PR DESCRIPTION
**Pull request by TheButlah**
image

(this is a joke this code is gonna suck to review because its so complicated and cursed, I'm so sorry @tomthecarrot I tried to compartmentalize and simplify it where I could but its a fundamentally hard problem I think).

~~Marked as draft until we can test it using the other draft PRs.~~

Note: This PR actually breaks the old unit tests so they dont compile, that is fixed in #149. So read and review this pr, but actually test with the other PR.

### CODE REVIEW

File: client/serialize/rust/src/deserializer/old.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
+use bimap::BiHashMap;
```
Comment (TheButlah) [inline client/serialize/rust/src/deserializer/old.rs:1]: This file was moved from the toplevel deserializer module and renamed to `old`. I reused some of this code in the second attempt at implementing the deserializer. So you can ignore this file - just review the stuff in the `new` module, including the reused old code.

File: client/serialize/rust/src/deserializer/new/objects.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+use bimap::BiHashMap;
+
+use crate::{rs, types::ObjectsIdx};
+
+#[derive(Default)]
+pub struct InstantiatedObjects(BiHashMap<rs::ObjectHandle, ObjectsIdx>);
```
Comment (TheButlah) [inline client/serialize/rust/src/deserializer/new/objects.rs:7]: Need to add docstring

File: client/serialize/rust/src/deserializer/new/mod.rs [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+        // `rs::StateHandle` for the state until after we construct the object, so this
+        // will be used after object construction to re-associate these `StatesIdx`
+        // with the `rs::StateHandle`.
+        let mut null_states: Vec<(rs::StateId<rs::ObjectHandle>, StatesIdx)> = Vec::new();
+        for (state_id, obj_state_idx) in contract.state_iter().zip(obj_states.into_iter()) {
+            let obj_state_t = self.b.base_t.states().unwrap().get(obj_state_idx.0);
+            // Handle dynamic typing of union to access the property
+            use fb::TpPrimitive as P;
+
+            let prop = match obj_state_t.p_type() {
```
Comment (TheButlah) [inline client/serialize/rust/src/deserializer/new/mod.rs:376]: Add a 
```suggestion
            let prop: DynTpProperty = match obj_state_t.p_type() {
```

File: client/serialize/rust/src/deserializer/new/objects.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+impl InstantiatedObjects {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add_object(&mut self, handle: rs::ObjectHandle, idx: ObjectsIdx) {
+        assert!(matches!(
+            self.0.insert(handle, idx),
+            bimap::Overwritten::Neither
```
Comment (TheButlah) [inline client/serialize/rust/src/deserializer/new/objects.rs]: Add string panic message

**Review by tomthecarrot — APPROVED** [review]
💯

**Pull request merged**